### DISCUSSION TITLE
Accessing changeset in pimcore.dataobject.preUpdate

### DISCUSSION THREAD

**Discussion by cancan101** [pimcore/pimcore#17284] [Q&A / Support]
Inside the `pimcore.dataobject.preUpdate` handler, is there any way to access the "changeset", ie the set of fields that are changing and ideally what the before and after values are? 

For reference, the [doctrine `preUpdate` event _does_ provide that information](https://www.doctrine-project.org/projects/doctrine-orm/en/current/reference/events.html#preupdate):

>This means you have access to all the fields that have changed for this entity with their old and new value. The following methods are available on the PreUpdateEventArgs:
>getEntity() to get access to the actual entity.
getEntityChangeSet() to get a copy of the changeset array. Changes to this returned array do not affect updating.
hasChangedField($fieldName) to check if the given field name of the current entity changed.
getOldValue($fieldName) and getNewValue($fieldName) to access the values of a field.
setNewValue($fieldName, $value) to change the value of a field to be updated.

See also the [reference in the Symfony docs](https://symfony.com/doc/current/doctrine/events.html#doctrine-lifecycle-callbacks):
>Some lifecycle callbacks receive an argument that provides access to useful information such as the current entity manager (e.g. the preUpdate callback receives a PreUpdateEventArgs $event argument).

Does Pimcore expose this information in any way? Alternatively is there another lifecycle hook that should be used to access this information?

At the very least is there a way to see the list of fields that are "dirty" in the the `preUpdate` handler?

Related Q&A posted (no response): https://github.com/orgs/pimcore/discussions/17127.

**Comment by kingjia90** [comment]
To see the dirty fields, you could use [$this->object->getDirtyFields()](https://github.com/pimcore/pimcore/blob/b1bab678a43a82e500c05631bc2c2be192be85c5/models/Element/Traits/DirtyIndicatorTrait.php#L59) on the object.
https://github.com/pimcore/pimcore/blob/68f198d32add713f2e4025b5dcb6414754133145/lib/Event/Model/DataObjectEvent.php#L35

For the changeSet, i don't think we have anything yet in regards, but a potential idea would be to store the object version id (from which opened the object that is being updated) and in the event get it from the DB again and compare it against the dirty fields, or something like that?

**Reply by cancan101** [reply]
Something doesn't seem to be working correctly. If I modify any scalar/primitive fields (e.g. a text field), I am _not_ seeing that field listed in the `dirtyFields`. Changing a localized value or a relationship _does_ show up.
When modifying the scalar, the `dirtyFields` is `{"userModification":true,"modificationDate":true}`. If I modify a relationship field called `assets`, then I see: `{"userModification":true,"assets":true,"modificationDate":true}`

I am on Pimcore v10.6 so I had to access the `dirtyFields` using:
```php
$dirtyFields = (fn() => $this->o_dirtyFields)->call($product);
```

Looking at the generate code for my Object class (at least on v10), it does not look like setting a scalar fields updates the dirtyFields.

**Reply by markus-moser** [reply]
This kind of dirty detection is not supported by all data types (see https://github.com/pimcore/pimcore/blob/a96a60244d7d9bb26a0d69500a7075efc6fac686/models/DataObject/ClassDefinition/Data.php#L1302).

The only way for now to do this for those values is to get the original object out of the database and compare the attributes with the ones from the event.

**Reply by cancan101** [reply]
@markus-moser 

>The only way for now to do this for those values is to get the original object out of the database and compare the attributes with the ones from the event.

For a calculated field, is there _any_ way to see if the value is changing (i.e. upon re-calculation the new value is different from what it was before) or is the old value not stored in the database? I know calculated fields are in the query table, but will the value show up in the standard storage location allowing dirty calculation?

**Reply by jdreesen** [reply]
> The only way for now to do this for those values is to get the original object out of the database and compare the attributes with the ones from the event.

Which is of course pretty slow 😞
